### PR TITLE
FIR: check NESTED_CLASS_NOT_ALLOWED

### PR DESCRIPTION
**Pull request by tgeng**

### CODE REVIEW

File: idea/idea-frontend-fir/idea-fir-low-level-api/src/org/jetbrains/kotlin/idea/fir/low/level/api/file/structure/FileStructureElement.kt [hunk 1]
```diff
@@ -173,7 +174,10 @@ internal class NonReanalyzableDeclarationStructureElement(
                 firFile,
                 onDeclarationEnter = { firDeclaration ->
                     when {
-                        firDeclaration.isGeneratedDeclaration -> DiagnosticCollectorDeclarationAction.SKIP
+                        firDeclaration.isGeneratedDeclaration &&
```
Comment (punzki) [inline idea/idea-frontend-fir/idea-fir-low-level-api/src/org/jetbrains/kotlin/idea/fir/low/level/api/file/structure/FileStructureElement.kt:177]: Can we just use `SKIP_CURRENT_DECLARATION_AND_CHECK_NESTED` here? I believe this will let us skip the generated declaration itself (e.g., the anonymous object) but get to the real declarations inside (e.g., the nested class in the enum entry).

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import org.jetbrains.kotlin.descriptors.ClassKind
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors.NESTED_CLASS_NOT_ALLOWED
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.declarations.*
+
+object FirNestedClassChecker : FirRegularClassChecker() {
+    override fun check(declaration: FirRegularClass, context: CheckerContext, reporter: DiagnosticReporter) {
+        // Local enums / objects / companion objects are handled in different checks
```
Comment (punzki) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt:17]: Are these already currently handled in different checkers?

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+        when (containingDeclaration) {
+            is FirRegularClass -> {
+                if (!declaration.isInner && (containingDeclaration.isInner || containingDeclaration.isLocal)) {
+                    reporter.reportOn(declaration.source, NESTED_CLASS_NOT_ALLOWED, declaration.classKind.description, context)
+                }
+            }
+            is FirClass<*> -> {
+                // Since 1.3, enum entries can contain inner classes only.
+                // Companion objects are reported in with code WRONG_MODIFIER_CONTAINING_DECLARATION instead
```
Comment (punzki) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt:29]: Typo: should be "...reported with code..."

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt (hunk 2 above)
Comment (tgeng) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt:17]: TBH, I am not sure what the current status is. I am copying the existing logic in `compiler/frontend/src/org/jetbrains/kotlin/resolve/ModifiersChecker.java`. Let me rephrase the comment to say that local enums/objects/companion objects are handled with different diagnostic codes.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+package org.jetbrains.kotlin.fir.analysis.checkers.declaration
+
+import org.jetbrains.kotlin.descriptors.ClassKind
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors.NESTED_CLASS_NOT_ALLOWED
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.declarations.*
+
+object FirNestedClassChecker : FirRegularClassChecker() {
```
Comment (jsjeon) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt:15]: According to https://github.com/JetBrains/kotlin/blob/master/compiler/frontend/src/org/jetbrains/kotlin/resolve/ModifiersChecker.java#L188, this is applied to both regular class and anonymous object. Thus, this should extend `FirClassChecker` (and added to `classCheckers`)

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    override fun check(declaration: FirRegularClass, context: CheckerContext, reporter: DiagnosticReporter) {
+        // Local enums / objects / companion objects are handled with different diagnostic codes.
+        if (declaration.classKind.isSingleton && declaration.isLocal) return
```
Comment (jsjeon) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt:18]: According to https://github.com/JetBrains/kotlin/blob/master/compiler/frontend/src/org/jetbrains/kotlin/resolve/ModifiersChecker.java#L201, enum class, object, and companion object are bailed out here.

`isSingleton` checks if the class kind is enum entry or object. Here, we accidentally bail out for enum entry as well. You better explicitly check if the class kind is `ENUM_CLASS` or `OBJECT`.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    private val ClassKind.description: String
+        get() = when (this) {
+            ClassKind.CLASS -> "Class"
+            ClassKind.INTERFACE -> "Interface"
+            ClassKind.ENUM_CLASS -> "Enum class"
+            ClassKind.ENUM_ENTRY -> "Enum entry"
+            ClassKind.ANNOTATION_CLASS -> "Annotation class"
+            ClassKind.OBJECT -> "Object"
```
Comment (jsjeon) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt:44]: Shouldn't we differentiate "Object", "Companion Object", and "Anonymous Object"? You will need more than just `ClassKind` here.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDefaultErrorMessages.kt [hunk 7]
```diff
@@ -208,6 +209,7 @@ class FirDefaultErrorMessages : DefaultErrorMessages.Extension {
             map.put(VARIABLE_EXPECTED, "Variable expected")
             map.put(RETURN_NOT_ALLOWED, "'return' is not allowed here")
             map.put(DELEGATION_IN_INTERFACE, "Interfaces cannot use delegation")
+            map.put(NESTED_CLASS_NOT_ALLOWED, "{0} is not allowed here", NULLABLE_STRING)
```
Comment (jsjeon) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDefaultErrorMessages.kt:212]: Are we going to pass nullable string? Why not `TO_STRING`?

Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDefaultErrorMessages.kt:212]: Oh, I guess you wanted to nail down to `String` type at least. Either should work, then. `TO_STRING` will weight more on nullability. :)

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt (hunk 4 above)
Comment (tgeng) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt:15]: It seems line 198 in ModifiersChecker.java has ruled out anonymous objects. Does it not mean what I think it means?

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt (hunk 5 above)
Comment (tgeng) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt:18]: enum entry bails out on line 198 in ModifiersChecker.java. So this should be equivalent.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt (hunk 6 above)
Comment (tgeng) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt:44]: IIUC, companioin object and anonymous object should have bailed out earlier. So we never need to differentiate them.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDefaultErrorMessages.kt (hunk 7 above)
Comment (tgeng) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDefaultErrorMessages.kt:212]: Yeah I wasn't sure what to use. I wish there is an additional `STRING` option :S

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt (hunk 4 above)
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt:15]: Cool, seems right. Let's add comments then, like we don't need to visit anonymous object and enum entry.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt (hunk 5 above)
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt:18]: Okay, using `isSingleton` seems fine in the sense that it can help us bail out for local object.

But, still, you won't need to check if the class kind is `ENUM_CLASS` and bail out too? Note that enum class and enum entries are different.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt (hunk 6 above)
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt:44]: Oh, so, they won't reach here technically. Could you comments to explicitly mention that?

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt (hunk 5 above)
Comment (tgeng) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt:18]: Ah thanks! I missed that.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt (hunk 6 above)
Comment (tgeng) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt:44]: Actually companion object is still needed! Let me add a comment for anonymous object.

File: compiler/fir/raw-fir/psi2fir/src/org/jetbrains/kotlin/fir/builder/RawFirBuilder.kt [hunk 8]
```diff
@@ -793,7 +793,10 @@ class RawFirBuilder(
         override fun visitClassOrObject(classOrObject: KtClassOrObject, data: Unit): FirElement {
             return withChildClassName(
                 classOrObject.nameAsSafeName,
-                classOrObject.isLocal || classOrObject.getStrictParentOfType<KtEnumEntry>() != null
+                classOrObject.isLocal
+                        // TODO: currently enum entry initializer is represented in FIR as an FirAnonymousObject. Because of this, all
+                        //  nested declarations are now marked local. This causes the FirNestedClassChecker to ignore some invalid programs.
```
Comment (jsjeon) [inline compiler/fir/raw-fir/psi2fir/src/org/jetbrains/kotlin/fir/builder/RawFirBuilder.kt:798]: Nit: perhaps point to https://youtrack.jetbrains.com/issue/KT-45115 ?

**Review by mglukhikh — APPROVED** [review]

**Comment by mglukhikh** [comment]
Merged manually 👍
